libcontainer/rootfs_linux.go

```diff
@@ -377,25 +377,35 @@ func mountToRootfs(m *configs.Mount, c *mountConfig) error {
 	rootfs := c.root
 	mountLabel := c.label
 	mountFd := c.fd
-	dest, err := securejoin.SecureJoin(rootfs, m.Destination)
-	if err != nil {
-		return err
-	}
 
 	switch m.Device {
 	case "proc", "sysfs":
 		// If the destination already exists and is not a directory, we bail
-		// out This is to avoid mounting through a symlink or similar -- which
+		// out. This is to avoid mounting through a symlink or similar -- which
 		// has been a "fun" attack scenario in the past.
 		// TODO: This won't be necessary once we switch to libpathrs and we can
 		//       stop all of these symlink-exchange attacks.
+		dest := m.Destination
+		if !strings.HasPrefix(dest, rootfs) {
+			// Do not use securejoin as it resolves symlinks.
+			dest = filepath.Join(rootfs, dest)
+		}
 		if fi, err := os.Lstat(dest); err != nil {
 			if !os.IsNotExist(err) {
 				return err
 			}
-		} else if fi.Mode()&os.ModeDir == 0 {
+		} else if !fi.IsDir() {
 			return fmt.Errorf("filesystem %q must be mounted on ordinary directory", m.Device)
 		}
+	}
+
+	dest, err := securejoin.SecureJoin(rootfs, m.Destination)
+	if err != nil {
+		return err
+	}
+
+	switch m.Device {
+	case "proc", "sysfs":
```

Comment:
This can be simplified by just having the proc/sysfs case be a standalone `switch` which returns and then have everything afterwards. Something like:

```patch
diff --git a/libcontainer/rootfs_linux.go b/libcontainer/rootfs_linux.go
index 2a98372b561d..b698718da85d 100644
--- a/libcontainer/rootfs_linux.go
+++ b/libcontainer/rootfs_linux.go
@@ -377,13 +377,18 @@ func mountToRootfs(m *configs.Mount, c *mountConfig) error {
        rootfs := c.root
        mountLabel := c.label
        mountFd := c.fd
-       dest, err := securejoin.SecureJoin(rootfs, m.Destination)
-       if err != nil {
-               return err
-       }

+       // procfs and sysfs are special because we need to ensure they are actually
+       // mounted on a specific path in a container without any funny business.
        switch m.Device {
        case "proc", "sysfs":
+               // Scope the path to the rootfs. We cannot use SecureJoin because it
+               // resolves symlinks which means we cannot check if the target is a
+               // symlink.
+               dest := m.Destination
+               if !strings.HasPrefix(dest, rootfs) {
+                       dest = filepath.Join(rootfs, dest)
+               }
                // If the destination already exists and is not a directory, we bail
                // out This is to avoid mounting through a symlink or similar -- which
                // has been a "fun" attack scenario in the past.
@@ -393,7 +398,7 @@ func mountToRootfs(m *configs.Mount, c *mountConfig) error {
                        if !os.IsNotExist(err) {
                                return err
                        }
-               } else if fi.Mode()&os.ModeDir == 0 {
+               } else if !fi.IsDir() {
                        return fmt.Errorf("filesystem %q must be mounted on ordinary directory", m.Device)
                }
                if err := os.MkdirAll(dest, 0o755); err != nil {
@@ -401,6 +406,14 @@ func mountToRootfs(m *configs.Mount, c *mountConfig) error {
                }
                // Selinux kernels do not support labeling of /proc or /sys
                return mountPropagate(m, rootfs, "", nil)
+       }
+
+       dest, err := securejoin.SecureJoin(rootfs, m.Destination)
+       if err != nil {
+               return err
+       }
+
+       switch m.Device {
        case "mqueue":
                if err := os.MkdirAll(dest, 0o755); err != nil {
                        return err
```

Reply:
updated